### DISCUSSION TITLE
Cortex-A9 fails to build cmsis due to missing core_ca.h

### DISCUSSION THREAD

**Discussion by yroeht** [zephyrproject-rtos/zephyr#49815] [Q&A]
**Describe the bug**
I wanted to check out the Cortex-A9 support, so I tried building the hello_world sample. This issue seems unrelated to the specific sample though. It is also probably unrelated to the the target and entirely due to user error wrt. modules.

The build fails due to a missing header file, which I believe is located in cmsis module (modules/hal/cmsis) at ./hal/cmsis/CMSIS/Core_A/Include/core_ca.h

I have no experience with Zephyr modules, so I may be doing something wrong here. I tried using `-DZEPHYR_MODULES` and `-DZEPHYR_EXTRA_MODULES` to no avail.

**To Reproduce**
1. west init -m https://github.com/zephyrproject-rtos/zephyr --mr v3.1.0 zephyrproject
2. cd zephyrproject
3. west update
4. . west build -b qemu_cortex_a9 zephyr/samples/hello_world -- -DZEPHYR_TOOLCHAIN_VARIANT=arm-none-eabi
5. . See error

**Expected behavior**
I expect a successful compilation of the hello_world sample

**Impact**
This is stopping me (blocker) from implementing support for another Cortex-A aarch32 soc for my work.

**Logs and console output**
```
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world
-- Cache files will be written to: /home/tranquet/.cache/zephyr
-- Zephyr version: 3.1.0 (/home/tranquet/zephyr)
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- Board: qemu_cortex_a9
-- Found dtc: /usr/local/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /home/tranquet/zephyr/boards/arm/qemu_cortex_a9/qemu_cortex_a9.dts
-- Generated zephyr.dts: /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/zephyr.dts                                                                                             
-- Generated devicetree_unfixed.h: /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/include/generated/devicetree_unfixed.h                                                       
-- Generated device_extern.h: /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/include/generated/device_extern.h                                                                 
-- Including generated dts.cmake file: /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/dts.cmake                                                                                
/home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/zephyr.dts:127.4-11: Warning (ranges_format): /soc/gpio@e000a000:ranges: empty "ranges" property but its #size-cells (0) differs
from /soc (1)
Parsing /home/tranquet/zephyr/Kconfig
Loaded configuration '/home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/.config'                                                                                                  
No change to configuration in '/home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/.config'                                                                                         
No change to Kconfig header in '/home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/include/generated/autoconf.h'                                                                   
-- Configuring done
-- Generating done
-- Build files have been written to: /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build                                                                                                   
-- west build: building application
[2/134] Building C object zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj
FAILED: zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj
/opt/soft/toolchains64/arm-none-eabi/bin/arm-none-eabi-gcc -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I/home/tranquet/zephyr/kernel/include -I/home/tranquet/zephyr/arch/arm/include -I/home/tranquet/zephyr/include/zephyr -I/home/tranquet/zephyr/include -I/home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/include/generated -I/home/tranquet/zephyr/soc/arm/xilinx_zynq7000/xc7zxxxs -isystem
/home/tranquet/zephyr/lib/libc/minimal/include -isystem /opt/soft/toolchains64/timestamped/gcc-arm-10.2-2020.11-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/10.2.1/include -isystem /opt/soft/toolchains64/timestamped/gcc-arm-10.2-2020.11-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/10.2.1/include-fixed -Os -imacros /home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a9 -mabi=aapcs -mfp16-format=ieee -imacros /home/tranquet/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/tranquet/nobackup/GIT/zephyrproject/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/tranquet/zephyr=ZEPHYR_BASE -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -c /home/tranquet/zephyr/arch/arm/core/offsets/offsets.c
In file included from /home/tranquet/zephyr/include/zephyr/arch/arm/aarch32/cortex_a_r/sys_io.h:19,
                 from /home/tranquet/zephyr/include/zephyr/arch/arm/aarch32/arch.h:42,
                 from /home/tranquet/zephyr/include/zephyr/arch/cpu.h:19,
                 from /home/tranquet/zephyr/include/zephyr/kernel_includes.h:33,
                 from /home/tranquet/zephyr/include/zephyr/kernel.h:17,
                 from /home/tranquet/zephyr/arch/arm/core/offsets/offsets_aarch32.c:28,
                 from /home/tranquet/zephyr/arch/arm/core/offsets/offsets.c:9:
/home/tranquet/zephyr/include/zephyr/arch/arm/aarch32/cortex_a_r/cmsis.h:59:10: fatal error: core_ca.h: No such file or directory                                                                                 
   59 | #include <core_ca.h>
      |          ^~~~~~~~~~~
compilation terminated.
```

**Environment:**
 - OS: GNU/Linux Debian 10
 - Toolchain arm-none-eabi
 - v3.1.0

Thank you for your time and sorry for the inconvenience.

**Comment by stephanosio** [comment]
This is working for me with both GNU Arm Embedded and Zephyr SDK.

Something must be wrong with your environment.

Also `ZEPHYR_TOOLCHAIN_VARIANT=arm-none-eabi` is simply wrong. Please see https://docs.zephyrproject.org/latest/develop/toolchains/gnu_arm_embedded.html

**Reply by yroeht** [reply]
> Also `ZEPHYR_TOOLCHAIN_VARIANT=arm-none-eabi` is simply wrong.

I don't understand how it is wrong, its just another arm toolchain, but since it works for you with Zephyr SDK I will give it a try.



I have extracted ~/zephyr-sdk-0.14.2

Is the proper way to proceed to `cp -r ~/zephyr-sdk-0.14.2/arm-zephyr-eabi ~/nobackup/GIT/zephyrproject/zephyr/cmake/toolchain/` ? And then redo the previously mentioned steps, with s/arm-none-eabi/arm-zephyr-eabi/ ?

i.e.

1.  west init -m https://github.com/zephyrproject-rtos/zephyr --mr v3.1.0 zephyrproject
2.  cd zephyrproject
3. west update
4. west build -b qemu_cortex_a9 samples/hello_world -- -DZEPHYR_TOOLCHAIN_VARIANT=arm-**zephyr**-eabi
5. See error

or Should I set -D{TOOLCHAIN}_TOOLCHAIN_PATH? But then what value do I use for TOOLCHAIN?


I did notice though

-- Application: /home/tranquet/nobackup/GIT/**zephyrproject/zephyr**/samples/hello_world
-- Cache files will be written to: /home/tranquet/.cache/zephyr
-- Zephyr version: 3.1.0 (/home/tranquet/**zephyr**)

It seems to be using an other (previous, non-West) Zephyr, even though

```
$ echo $ZEPHYR_BASE
/home/tranquet/nobackup/GIT/zephyrproject/zephyr
```

Why is it using that other Zephyr in my home?

**Reply by stephanosio** [reply]
> I don't understand how it is wrong, its just another arm toolchain

Please read https://docs.zephyrproject.org/latest/develop/toolchains/gnu_arm_embedded.html

**Reply by stephanosio** [reply]
Also it would be good to read the Getting Started Guide:
https://docs.zephyrproject.org/latest/develop/getting_started/index.html#install-zephyr-sdk

**Reply by stephanosio** [reply]
Also https://docs.zephyrproject.org/latest/develop/toolchains/zephyr_sdk.html on how to manually configure environment variables to use Zephyr SDK.

**Reply by yroeht** [reply]
> > I don't understand how it is wrong, its just another arm toolchain
> 
> Please read https://docs.zephyrproject.org/latest/develop/toolchains/gnu_arm_embedded.html

Done

> Also it would be good to read the Getting Started Guide: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#install-zephyr-sdk

Done

> Also https://docs.zephyrproject.org/latest/develop/toolchains/zephyr_sdk.html on how to manually configure environment variables to use Zephyr SDK.

Done

Now what?

**Reply by yroeht** [reply]
**I got it to work.**

1. #install Zephyr SDK
2. wget https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.15.0/zephyr-sdk-0.15.0_linux-x86_64.tar.gz
3. wget -O - https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.15.0/sha256.sum | shasum --check --ignore-missing
4. tar xvf zephyr-sdk-0.15.0_linux-x86_64.tar.g
5. cd zephyr-sdk-0.15.0
6. ./setup.sh
7. cd
8. #grab Zephyr and compile sample
9. west init -m https://github.com/zephyrproject-rtos/zephyr --mr v3.1.0 zephyrproject
10. cd zephyrproject
11. west update
12. export ZEPHYR_BASE=~/zephyrproject/zephyr
13. west build -b qemu_cortex_a9 zephyr/samples/hello_world

Not entirely sure **which** toolchain is being used though:
```
-- Found dtc: /home/tranquet/zephyr-sdk-0.14.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6")
...
-- Found assembler: **/opt/soft/toolchains64/arm-none-eabi/bin/arm-none-eabi-gcc**
```

Setting ZEPHYR_TOOLCHAIN_VARIANT and ZEPHYR_SDK_INSTALL_DIR or not seems to have no effect. Based on https://docs.zephyrproject.org/latest/develop/beyond-GSG.html#gs-toolchain-update (which was not one of the docs recommended), the value for ZEPHYR_TOOLCHAIN_VARIANT should be `zephyr`:
>If the [ZEPHYR_TOOLCHAIN_VARIANT](https://docs.zephyrproject.org/latest/develop/env_vars.html#envvar-ZEPHYR_TOOLCHAIN_VARIANT) environment variable is not zephyr, then either unset it or change its value to zephyr to make sure Zephyr SDK is selected.

I have been trying to reproduce my issue with no success. I am certain the SDK had already been installed though so  that's not it.
